android/src/com/unciv/app/CustomSaveLocationHelperAndroid.kt

```diff
@@ -0,0 +1,115 @@
+package com.unciv.app
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import android.os.Build
+import androidx.annotation.GuardedBy
+import androidx.annotation.RequiresApi
+import com.unciv.logic.CustomSaveLocationHelper
+import com.unciv.logic.GameInfo
+import com.unciv.logic.GameSaver
+import com.unciv.logic.GameSaver.json
+
+// The Storage Access Framework is available from API 19 and up:
+// https://developer.android.com/guide/topics/providers/document-provider
+@RequiresApi(Build.VERSION_CODES.KITKAT)
+class CustomSaveLocationHelperAndroid(private val activity: Activity) : CustomSaveLocationHelper {
+    // This looks a little scary but it's really not so bad. Whenever a load or save operation is
+    // attempted, the game automatically autosaves as well (but on a separate thread), so we end up
+    // with a race condition when trying to handle both operations in parallel. In order to work
+    // around that, the callbacks are given an arbitrary index beginning at 100 and incrementing
+    // each time, and this index is used as the requestCode for the call to startActivityForResult()
+    // so that we can map it back to the corresponding callback when onActivityResult is called
+    @GuardedBy("this")
+    @Volatile
+    private var callbackIndex = 100
+    @GuardedBy("this")
+    private val callbacks = ArrayList<IndexedCallback>()
+
+    override fun saveGame(gameInfo: GameInfo, gameName: String, forcePrompt: Boolean, block: (() -> Unit)?) {
+        val callbackIndex = synchronized(this) {
+            val index = callbackIndex++
+            callbacks.add(IndexedCallback(
+                    index,
+                    { uri ->
+                        uri?.let {
+                            saveGame(gameInfo, it)
+                        }
+                        block?.invoke()
+                    }
+            ))
+            index
+        }
+        gameInfo.customSaveLocation?.let { customSaveLocation ->
```

Comment:
The nullability check can be merged into the if(!forcePrompt) check for better readability